SmartDeviceLink/private/SDLVoiceCommandUpdateOperation.m

```diff
@@ -0,0 +1,225 @@
+- (void)sdl_sendDeleteCurrentVoiceCommands:(void(^)(void))completionHandler {
+    if (self.oldVoiceCommands.count == 0) {
+        SDLLogD(@"No voice commands to delete");
+        return completionHandler();
+    }
+
+    NSArray<SDLDeleteCommand *> *deleteVoiceCommands = [self sdl_deleteCommandsForVoiceCommands:self.oldVoiceCommands];
+    __block NSMutableDictionary<SDLDeleteCommand *, NSError *> *errors = [NSMutableDictionary dictionary];
+    __weak typeof(self) weakSelf = self;
+    [self.connectionManager sendRequests:deleteVoiceCommands progressHandler:^(__kindof SDLRPCRequest * _Nonnull request, __kindof SDLRPCResponse * _Nullable response, NSError * _Nullable error, float percentComplete) {
+        // Add the request to the error dict or remove it from the current voice commands array
+        if (error != nil) {
+            errors[request] = error;
+        } else {
+            [weakSelf sdl_removeCurrentVoiceCommandWithCommandId:((SDLDeleteCommand *)request).cmdID.unsignedIntValue];
+        }
+
+        SDLLogV(@"Deleting voice commands progress: %f", percentComplete);
+    } completionHandler:^(BOOL success) {
+        if (!success) {
+            SDLLogE(@"Error deleting old voice commands: %@", errors);
+            weakSelf.internalError = [NSError sdl_menuManager_failedToUpdateWithDictionary:errors];
+            return completionHandler();
+        }
+
+        SDLLogD(@"Finished deleting old voice commands");
+        return completionHandler();
+    }];
+}
+
+/// Send AddCommand RPCs for voice commands that should be added
+/// @param completionHandler A handler called when all AddCommands have completed
+- (void)sdl_sendCurrentVoiceCommands:(void(^)(void))completionHandler {
+    if (self.pendingVoiceCommands.count == 0) {
+        SDLLogD(@"No voice commands to send");
+        return completionHandler();
+    }
+
+    NSArray<SDLAddCommand *> *addCommandsToSend = [self sdl_addCommandsForVoiceCommands:self.pendingVoiceCommands];
+    __block NSMutableDictionary<SDLAddCommand *, NSError *> *errors = [NSMutableDictionary dictionary];
+    __weak typeof(self) weakSelf = self;
+    [self.connectionManager sendRequests:addCommandsToSend progressHandler:^(__kindof SDLRPCRequest * _Nonnull request, __kindof SDLRPCResponse * _Nullable response, NSError * _Nullable error, float percentComplete) {
+        // Add the request to the error dict or add it to the current voice commands array
+        if (error != nil) {
+            errors[request] = error;
+        } else {
+            [weakSelf.currentVoiceCommands addObject:[weakSelf sdl_pendingVoiceCommandWithCommandId:((SDLAddCommand *)request).cmdID.unsignedIntValue]];
+        }
+
+        SDLLogV(@"Sending voice commands progress: %f", percentComplete);
+    } completionHandler:^(BOOL success) {
+        if (!success) {
+            SDLLogE(@"Failed to send main menu commands: %@", errors);
```

Comment:
It should have been doing that. After investigation, this seems to be some sort of Xcode bug. The values are not actually `null`. If you take the dictionary and do an `allValues`, you'll see the actual errors. I rewrote the logs to hopefully make this work, even though it will be less clear.